src/webgpu/shader/execution/shader_io/compute_builtins.spec.ts

```diff
@@ -0,0 +1,286 @@
+[[group(0), binding(2)]] var<storage, write> global_id_out : V;
+[[group(0), binding(3)]] var<storage, write> group_id_out : V;
+[[group(0), binding(4)]] var<storage, write> num_groups_out : V;
+
+${structures}
+
+let group_width = ${t.params.groupSize.x}u;
+let group_height = ${t.params.groupSize.y}u;
+let group_depth = ${t.params.groupSize.z}u;
+
+[[stage(compute), workgroup_size(group_width, group_height, group_depth)]]
+fn main(
+  ${params}
+  ) {
+  let group_index = ((${group_id}.z * ${num_groups}.y) + ${group_id}.y) * ${num_groups}.x + ${group_id}.x;
+  let global_index = group_index * ${invocationsPerGroup}u + ${local_index};
+  local_id_out.data[global_index] = ${local_id};
+  local_index_out.data[global_index] = ${local_index};
+  global_id_out.data[global_index] = ${global_id};
+  group_id_out.data[global_index] = ${group_id};
+  num_groups_out.data[global_index] = ${num_groups};
+}
+`;
+
+    const pipeline = t.device.createComputePipeline({
+      compute: {
+        module: t.device.createShaderModule({
+          code: wgsl,
+        }),
+        entryPoint: 'main',
+      },
+    });
+
+    // Helper to create a `size`-byte buffer with binding number `binding`.
+    function createBuffer(size: number, binding: number) {
+      const buffer = t.device.createBuffer({
+        size,
+        usage: GPUBufferUsage.STORAGE | GPUBufferUsage.COPY_SRC,
+      });
+      t.trackForCleanup(buffer);
+
+      bindGroupEntries.push({
+        binding,
+        resource: {
+          buffer,
+        },
+      });
+
+      return buffer;
+    }
+
+    // Create the output buffers.
+    const bindGroupEntries: GPUBindGroupEntry[] = [];
+    const localIdBuffer = createBuffer(totalInvocations * 16, 0);
+    const localIndexBuffer = createBuffer(totalInvocations * 4, 1);
+    const globalIdBuffer = createBuffer(totalInvocations * 16, 2);
+    const groupIdBuffer = createBuffer(totalInvocations * 16, 3);
+    const numGroupsBuffer = createBuffer(totalInvocations * 16, 4);
+
+    const bindGroup = t.device.createBindGroup({
+      layout: pipeline.getBindGroupLayout(0),
+      entries: bindGroupEntries,
+    });
+
+    // Run the shader.
+    const encoder = t.device.createCommandEncoder();
+    const pass = encoder.beginComputePass();
+    pass.setPipeline(pipeline);
+    pass.setBindGroup(0, bindGroup);
+    pass.dispatch(t.params.numGroups.x, t.params.numGroups.y, t.params.numGroups.z);
+    pass.endPass();
+    t.queue.submit([encoder.finish()]);
+
+    type vec3 = { x: number; y: number; z: number };
+
+    // Helper to check that the vec3<u32> value at each index of the provided `output` buffer
+    // matches the expected value for that invocation, as generated by the `getBuiltinValue`
+    // function. The `name` parameter is the builtin name, used for error messages.
+    const checkEachIndex = (
+      output: Uint32Array,
+      name: string,
+      getBuiltinValue: (groupId: vec3, localId: vec3) => vec3
+    ) => {
```

Comment:
This looks good, though I wonder if it could be done with the `checkElementsEqualGenerated` helper in `check_contents.ts`?

If it's not straightforward don't worry about it. It may require recomputing gx/gy/gz/lx/ly/lz from the buffer index which would probably be a pain.

Reply:
Yeah I initially tried this, but recomputing the group/local indices from the buffer index was awkward, and also I'm not sure how to handle the padding bytes since these buffers are arrays of `vec3` with a 16 byte stride.